src/app-layer-ftp.c

```diff
@@ -54,11 +54,136 @@
 #include "util-memcmp.h"
 #include "util-memrchr.h"
 #include "util-byte.h"
+#include "util-mem.h"
+#include "util-misc.h"
 
 #ifdef HAVE_RUST
 #include "rust-ftp-mod-gen.h"
 #endif
 
+uint64_t ftp_config_memcap = 0;
+
+SC_ATOMIC_DECLARE(uint64_t, ftp_memuse);
+SC_ATOMIC_DECLARE(uint64_t, ftp_memcap);
+
+static void FTPParseMemcap(void)
+{
+    const char *conf_val;
+
+    /** set config values for memcap, prealloc and hash_size */
+    if ((ConfGet("app-layer.protocols.ftp.memcap", &conf_val)) == 1)
+    {
+        if (ParseSizeStringU64(conf_val, &ftp_config_memcap) < 0) {
+            SCLogError(SC_ERR_SIZE_PARSE, "Error parsing ftp.memcap "
+                       "from conf file - %s.  Killing engine",
+                       conf_val);
+            exit(EXIT_FAILURE);
+        }
+        SCLogInfo("FTP memcap: %"PRIu64, ftp_config_memcap);
+    } else {
+        /* default to unlimited */
+        ftp_config_memcap = 0;
+    }
+
+    SC_ATOMIC_INIT(ftp_memuse);
+    SC_ATOMIC_INIT(ftp_memcap);
+}
+
+static void FTPIncrMemuse(uint64_t size)
+{
+    (void) SC_ATOMIC_ADD(ftp_memuse, size);
+    return;
+}
+
+static void FTPDecrMemuse(uint64_t size)
+{
+    (void) SC_ATOMIC_SUB(ftp_memuse, size);
+    return;
+}
+
+uint64_t FTPMemuseGlobalCounter(void)
+{
+    uint64_t tmpval = SC_ATOMIC_GET(ftp_memuse);
+    return tmpval;
+}
+
+uint64_t FTPMemcapGlobalCounter(void)
+{
+    uint64_t tmpval = SC_ATOMIC_GET(ftp_memcap);
+    return tmpval;
+}
+
+/**
+ *  \brief Check if alloc'ing "size" would mean we're over memcap
+ *
+ *  \retval 1 if in bounds
+ *  \retval 0 if not in bounds
+ */
+static int FTPCheckMemcap(uint64_t size)
+{
+    if (ftp_config_memcap == 0 || size + SC_ATOMIC_GET(ftp_memuse) <= ftp_config_memcap)
+        return 1;
+    (void) SC_ATOMIC_ADD(ftp_memcap, 1);
+    return 0;
+}
+
+static void *FTPMalloc(size_t size)
+{
+    void *ptr = NULL;
+
+    if (FTPCheckMemcap((uint32_t)size) == 0)
+        return NULL;
+
+    ptr = SCMalloc(size);
+
+    if (unlikely(ptr == NULL))
+        return NULL;
+
+    FTPIncrMemuse((uint64_t)size);
+
+    return ptr;
+}
+
+static void *FTPCalloc(size_t n, size_t size)
+{
+    void *ptr = NULL;
+
+    if (FTPCheckMemcap((uint32_t)(n * size)) == 0)
+        return NULL;
+
+    ptr = SCCalloc(n, size);
+
+    if (unlikely(ptr == NULL))
+        return NULL;
+
+    FTPIncrMemuse((uint64_t)(n * size));
+
+    return ptr;
+}
+
+static void *FTPRealloc(void *ptr, size_t orig_size, size_t size)
```

Comment:
looking at this it seems shrinking actually won't work correctly wrt memcaps. Same for the original HTPRealloc version.

Reply:
Indeed. Want me to fix it or do you take it ?

Reply:
Looks like ReassembleRealloc is a better example to copy. Can you take it for FTP?

Reply:
Yes, sure.